multicluster/controllers/multicluster/serviceexport_controller.go

```diff
@@ -18,21 +18,118 @@ package multicluster
 
 import (
 	"context"
+	"reflect"
+	"sort"
+	"strconv"
+	"strings"
+	"time"
 
+	corev1 "k8s.io/api/core/v1"
+	apiequality "k8s.io/apimachinery/pkg/api/equality"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/tools/cache"
 	"k8s.io/klog/v2"
 	ctrl "sigs.k8s.io/controller-runtime"
 	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
 	k8smcsv1alpha1 "sigs.k8s.io/mcs-api/pkg/apis/v1alpha1"
 
+	mcsv1alpha1 "antrea.io/antrea/multicluster/apis/multicluster/v1alpha1"
+	"antrea.io/antrea/multicluster/controllers/multicluster/common"
 	"antrea.io/antrea/multicluster/controllers/multicluster/core"
 )
 
-// ServiceExportReconciler reconciles a ServiceExport object
-type ServiceExportReconciler struct {
-	Client                  client.Client
-	Scheme                  *runtime.Scheme
-	RemoteCommonAreaManager *core.RemoteCommonAreaManager
+type (
+	svcInfo struct {
+		name       string
+		namespace  string
+		clusterIPs []string
+		ports      []corev1.ServicePort
+		svcType    string
+	}
+
+	epInfo struct {
+		name       string
+		namespace  string
+		addressIPs []string
+		ports      []corev1.EndpointPort
+		labels     map[string]string
+	}
+
+	// ServiceExportReconciler reconciles a ServiceExport object
+	ServiceExportReconciler struct {
+		Client                  client.Client
+		Scheme                  *runtime.Scheme
+		remoteCommonAreaManager *core.RemoteCommonAreaManager
+		installedSvcs           cache.Indexer
+		installedEps            cache.Indexer
+	}
+)
+
+const (
+	// cached indexer
+	svcIndexerByType = "service.by.type"
+	epIndexerByLabel = "endpoints.by.label"
+)
+
+var (
+	leaderNamespace string
```

Comment:
Could we avoid global variables? Can they be struct members?

Reply:
done, added a new clusterConfig.

Reply:
I do not mean we must create a new struct to hold these variables, but meant for OO better not to use global variables. I saw you change to pass clusterConfig as func arguments, but originally I was thinking can they be members of ServiceExportReconciler.

Reply:
done, added them as a field of ServiceExportReconciler